### DISCUSSION TITLE
How to use DataGrid's DragDrop.DragEnter and DragDrop.Drop in mvvm

### DISCUSSION THREAD

**Discussion by howzew** [AvaloniaUI/Avalonia#14191] [Q&A] [answered]
Use the following code in xaml:

`<DataGrid x:Name="dg" DragDrop.AllowDrop="True" DragDrop.DragEnter="{Binding DragEnterCommand}" DragDrop.Drop="{Binding DropCommand}"/>`

But the error message "Unable to resolve suitable regular or attached property Drop on type Avalonia.Base:Avalonia.Input.DragDrop" will be prompted.

I checked this post[Files Drag'n'Drop](https://github.com/AvaloniaUI/Avalonia/discussions/10602)，According to this method, the error message still exists and I don’t know how to get file information in ViewModel.

```
namespace Demo.ViewModels
{
    public partial class TestViewModel : ObservableObject
    {
	[RelayCommand]
        private void Drop()
        {
            //How to get the dragged file information in ViewModel?
        }
    }
}
```

**Comment by thevortexcloud** [comment]
From what I can tell Drag and drop works with events, not commands. As such this is impossible to do the way you want. You can do it indirectly though by having your view subscribe to the event and then pass the result to your view model via a command though.

**Comment by howzew** [comment]
> From what I can tell Drag and drop works with events, not commands. As such this is impossible to do the way you want. You can do it indirectly though by having your view subscribe to the event and then pass the result to your view model via a command though.

That doesn't seem to be more convenient than using the command directly.

**Reply by thevortexcloud** [reply]
It's not. But Avalonia does not support directly binding to events yet. It's a planned feature from memory though.

**Reply by howzew** [reply]
Do you have any examples for the method you mentioned above?Because drag and drop can't seem to bind commands

**Reply by thevortexcloud** [reply, marked as answer]
I think this should work using ReactiveUI. I have never touched the drag and drop stuff before though. It's also apparently [not implemented on Linux](https://github.com/AvaloniaUI/Avalonia/issues/5273) so I also can't test it. So there may be better ways to do it:


In your View codebehind:

```csharp
   public MainWindow() {
        InitializeComponent();

        this.AddHandler(DragDrop.DropEvent, Drop);
    }

    private async void Drop(object? sender, DragEventArgs e) {
        var files = e.Data.GetFileNames();
        await ((MainWindowViewModel)this.DataContext).DropCommand?.Execute(files);
    }
```


In your view model:
```csharp
    public ReactiveCommand<IEnumerable<string>, Unit> DropCommand { get; }

    public MainWindowViewModel() {
        DropCommand = ReactiveCommand.Create<IEnumerable<string>>(DropCommandHandler);
    }

    private void DropCommandHandler(IEnumerable<string> obj) {
        Console.WriteLine(string.Join(Environment.NewLine, obj));
    }
```

**Reply by howzew** [reply]
Thank you very much for your reply, but my entire project is using CommunityToolkit.Mvvm, and I have never used ReactiveUI, so, how should I write the above code if using CommunityToolkit.Mvvm?

**Reply by thevortexcloud** [reply]
I have never used it, but I imagine it should be mostly the same. The only thing I used from ReactiveUI in that example was the Command declaration. As such: based on what I am reading [here](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/relaycommand), you just need to change your command action to have an argument. EG
```csharp
	[RelayCommand]
        private void Drop(IEnumerable<string> files)
```

I think everything else is the same.

**Reply by thevortexcloud** [reply]
I forgot to mention, if you don't specifically need a command, your view can also just call a method directly on your view model. It still follows MVVM properly in that case.

**Reply by howzew** [reply]
Thank you so much. Your code is effective. I successfully implemented this function. I would like to express my gratitude!

**Reply by thevortexcloud** [reply]
You are welcome! Make sure you mark it as the answer so people can find it easily in future.
